File: src/lib/central/pty.ts
// Typed bridge to the embedded terminal (src-tauri/src/pty). Sessions live in
// the app process; a view attaches (history + cursor), then follows the live
// `pty://data` stream, acking each chunk once xterm has written it.

import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export type PtyForeground = { running: boolean; label: string; pid: number | null };

export type PtySessionInfo = {
  id: string;
  title: string;
  cwd: string | null;
  command: string[];
  pid: number | null;
  cols: number;
  rows: number;
  alive: boolean;
  exit_code: number | null;
  exit_signal: string | null;
  created_at: number;
  foreground: PtyForeground;
};

export type PtyOpenRequest = {
  id?: string;
  cwd?: string | null;
  /** argv; empty opens the default shell. */
  command?: string[];
  env?: Record<string, string>;
  cols?: number;
  rows?: number;
  title?: string;
  /** Typed after start, not submitted. */
  initial_input?: string;
  history_cap?: number;
};

export type PtyAttached = {
  info: PtySessionInfo | null;
  /** base64 of the sanitized history. */
  history: string;
  generation: number;
  /** Absolute offset the history ends at. */
  offset: number;
  alive: boolean;
};

export type PtyDataEvent = { id: string; generation: number; seq: number; start: number; data: string };
export type PtyExitEvent = { id: string; code: number | null; signal: string | null };
export type PtyActivityEvent = { id: string } & PtyForeground;

/** What "add selection to chat" hands the composer. */
export type TerminalSelection = { text: string; lines: [number, number]; terminal_id: string };

export const ptyOpen = (req: PtyOpenRequest) => invoke<PtySessionInfo>("pty_open", { req });
export const ptyWrite = (id: string, data: string) => invoke<void>("pty_write", { id, data });
export const ptyResize = (id: string, cols: number, rows: number) => invoke<void>("pty_resize", { id, cols, rows });
export const ptyClose = (id: string, deleteHistory = false) => invoke<void>("pty_close", { id, deleteHistory });
export const ptyList = () => invoke<PtySessionInfo[]>("pty_list");
export const ptyAttach = (id: string) => invoke<PtyAttached>("pty_attach", { id });
export const ptyDetach = (id: string) => invoke<void>("pty_detach", { id });
export const ptyAck = (id: string, seq: number) => invoke<void>("pty_ack", { id, seq });
export const ptyClear = (id: string) => invoke<number>("pty_clear", { id });
export const ptyForeground = (id: string) => invoke<PtyForeground>("pty_foreground", { id });

export function decodeBase64(s: string): Uint8Array {
  const bin = atob(s);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

export function errorCode(e: unknown): string {
  const s = String(e ?? "");
  const i = s.indexOf(":");
  return i > 0 ? s.slice(0, i) : s;
}

// One listener per event for the whole app, dispatched by terminal id.
export type PtyHandlers = {
  data?: (e: PtyDataEvent) => void;
  exit?: (e: PtyExitEvent) => void;
  activity?: (e: PtyActivityEvent) => void;
  stall?: () => void;
};

const subscribers = new Map<string, Set<PtyHandlers>>();
let hub: Promise<UnlistenFn[]> | null = null;

function dispatch<K extends keyof PtyHandlers>(id: string, key: K, arg: Parameters<NonNullable<PtyHandlers[K]>>[0]) {
  const set = subscribers.get(id);
  if (!set) return;
  for (const h of set) (h[key] as ((a: unknown) => void) | undefined)?.(arg);
}

function ensureHub(): Promise<UnlistenFn[]> {
  if (!hub) {
    hub = Promise.all([
      listen<PtyDataEvent>("pty://data", (e) => dispatch(e.payload.id, "data", e.payload)),
      listen<PtyExitEvent>("pty://exit", (e) => dispatch(e.payload.id, "exit", e.payload)),
      listen<PtyActivityEvent>("pty://activity", (e) => dispatch(e.payload.id, "activity", e.payload)),
      listen<{ id: string }>("pty://stall", (e) => dispatch(e.payload.id, "stall", undefined as never)),
    ]);
  }
  return hub;
}

/** Subscribes to one terminal's events. Resolves once the listeners are live,
 *  so an attach issued afterwards cannot miss a chunk. */
export async function subscribePty(id: string, handlers: PtyHandlers): Promise<() => void> {
  let set = subscribers.get(id);
  if (!set) {
    set = new Set();
    subscribers.set(id, set);
  }
  set.add(handlers);
  await ensureHub();
  return () => {
    const s = subscribers.get(id);
    if (!s) return;
    s.delete(handlers);
    if (s.size === 0) subscribers.delete(id);
  };
}

export function newTerminalId(): string {
  const r = typeof crypto !== "undefined" && "randomUUID" in crypto ? crypto.randomUUID() : `${Date.now()}-${Math.random().toString(36).slice(2)}`;
  return `term-${r}`;
}

File: src-tauri/src/commands/central/pty.rs
//! Tauri commands of the embedded terminal. The sessions live in
//! `crate::pty::manager()`; see that module for the event contract.

use tauri::AppHandle;

use crate::pty::{self, foreground::Foreground, Attached, OpenRequest, SessionInfo};

#[tauri::command]
pub async fn pty_open(app: AppHandle, req: OpenRequest) -> Result<SessionInfo, String> {
    let m = pty::manager();
    m.bind(&app);
    tauri::async_runtime::spawn_blocking(move || m.open(req))
        .await
        .map_err(|e| format!("PTY_OPEN: {e}"))?
}

#[tauri::command]
pub async fn pty_write(id: String, data: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || pty::manager().write(&id, &data))
        .await
        .map_err(|e| format!("PTY_WRITE: {e}"))?
}

#[tauri::command]
pub fn pty_resize(id: String, cols: u16, rows: u16) -> Result<(), String> {
    pty::manager().resize(&id, cols, rows)
}

#[tauri::command]
pub fn pty_close(id: String, delete_history: Option<bool>) -> Result<(), String> {
    pty::manager().close(&id, delete_history.unwrap_or(false))
}

#[tauri::command]
pub fn pty_list(app: AppHandle) -> Vec<SessionInfo> {
    let m = pty::manager();
    m.bind(&app);
    m.list()
}

#[tauri::command]
pub async fn pty_attach(app: AppHandle, id: String) -> Result<Attached, String> {
    let m = pty::manager();
    m.bind(&app);
    tauri::async_runtime::spawn_blocking(move || m.attach(&id))
        .await
        .map_err(|e| format!("PTY_ATTACH: {e}"))?
}

#[tauri::command]
pub fn pty_detach(id: String) -> Result<(), String> {
    pty::manager().detach(&id)
}

#[tauri::command]
pub fn pty_ack(id: String, seq: u64) -> Result<(), String> {
    pty::manager().ack(&id, seq)
}

#[tauri::command]
pub fn pty_clear(id: String) -> Result<u64, String> {
    pty::manager().clear(&id)
}

#[tauri::command]
pub async fn pty_foreground(id: String) -> Result<Foreground, String> {
    tauri::async_runtime::spawn_blocking(move || pty::manager().foreground(&id))
        .await
        .map_err(|e| format!("PTY_FOREGROUND: {e}"))?
}
